classes/task/update_object_tags.php

```
@@ -0,0 +1,85 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace tool_objectfs\task;
+
+use core\task\adhoc_task;
+use tool_objectfs\local\tag\tag_manager;
+
+/**
+ * Calculates and updates an objects tags in the external store.
+ *
+ * @package   tool_objectfs
+ * @author    Matthew Hilton <matthewhilton@catalyst-au.net>
+ * @copyright Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class update_object_tags extends adhoc_task {
+    /**
+     * Execute task
+     */
+    public function execute() {
+        if (!tag_manager::is_tagging_enabled_and_supported()) {
+            mtrace("Tagging feature not enabled or supported by filesystem, exiting.");
+            return;
+        }
+
+        // Since this adhoc task can requeue itself, ensure there is a fixed limit on the number
+        // of times this can happen, to avoid any accidental runaways.
+        $iterationlimit = get_config('tool_objectfs', 'maxtaggingiterations') ?: 0;
+        $iteration = !empty($this->get_custom_data()->iteration) ? $this->get_custom_data()->iteration : 0;
+
+        if (empty($iterationlimit) || empty($iteration)) {
+            mtrace("Invalid number of iterations, exiting.");
+            return;
+        }
+
+        if (abs($iteration) > abs($iterationlimit)) {
+            mtrace("Maximum number of iterations reached: " . $iteration . ", exiting.");
+            return;
+        }
```

Is there a reason why abs is used? I can't see any harm, but it's interesting to see this used over invalidating negatives.

---

No reason you are right through it should throw an exception, will tweak